finality-aleph/src/network.rs

```diff
@@ -404,16 +525,40 @@ impl<D: Clone + Encode + Decode, B: BlockT + 'static, N: Network<B> + Clone>
         }
     }
 
-    fn on_incoming_data(&self, session_id: SessionId, data: D) {
+    fn on_incoming_data(
+        &self,
+        session_id: SessionId,
+        index: MessageIndex,
+        data: D,
+        recipient: Recipient<NodeIndex>,
+    ) {
+        let message = DataMessage {
+            session_id,
+            index,
+            data: data.clone(),
+            recipient: recipient.clone(),
+        };
         let mut sessions = self.sessions.lock();
         if let Some(session_data) = sessions.get_mut(&session_id) {
             if session_data.status == SessionStatus::InProgress {
-                if let Err(e) = session_data.data_for_user.unbounded_send(data) {
-                    //TODO: need to write some logic on when an session should be terminated and make sure
-                    // that there are no issues with synchronization when terminating.
-                    session_data.status = SessionStatus::Terminated;
-                    debug!(target: "afa", "Error {:?} when passing a message event to session {:?}.", e, session_id);
+                if session_data.messages.contains(&message) {
+                    debug!(target: "afa", "Received data with old index in session {:?}.", session_id);
+                    return;
+                } else {
+                    session_data.messages.put(message, ());
                 }
+                if let Recipient::Target(node_id) = recipient {
+                    if node_id == session_data.auth_data.node_id {
+                        if let Err(e) = session_data.data_for_user.unbounded_send(data) {
```

Comment:
I might be wrong, but for Recipient::All we never place the message in the data_for_user channel?

Reply:
You might but you are not!